pkg/controller/machineset/machineset_controller.go

```diff
@@ -207,48 +296,353 @@ func (c *MachineSetController) handleErr(err error, key interface{}) {
+		return false
+	}
+	cluster, err := c.clusterForMachineSet(machineSet)
+	if err != nil {
+		return false
+	}
+	if !cluster.Status.Provisioned || cluster.Status.ProvisionedJobGeneration != cluster.Generation {
+		return false
+	}
+	switch machineSet.Spec.NodeType {
+	case clusteroperator.NodeTypeMaster:
+		return true
+	case clusteroperator.NodeTypeCompute:
+		masterMachineSet, err := c.machineSetsLister.MachineSets(machineSet.Namespace).Get(cluster.Status.MasterMachineSetName)
+		if err != nil {
+			return false
+		}
+		return masterMachineSet.Status.Installed && masterMachineSet.Status.InstalledJobGeneration == masterMachineSet.Generation
+	default:
+		return false
+	}
+}
+
+type jobOwnerControl struct {
+	controller *MachineSetController
+}
+
+func (c *jobOwnerControl) GetOwnerKey(owner metav1.Object) (string, error) {
+	return controller.KeyFunc(owner)
+}
+
+func (c *jobOwnerControl) GetOwner(namespace string, name string) (metav1.Object, error) {
+	return c.controller.machineSetsLister.MachineSets(namespace).Get(name)
+}
+
+func (c *jobOwnerControl) OnOwnedJobEvent(owner metav1.Object) {
+	machineSet, ok := owner.(*clusteroperator.MachineSet)
+	if !ok {
+		c.controller.logger.WithFields(log.Fields{"owner": owner.GetName(), "namespace": owner.GetNamespace()}).
+			Errorf("attempt to enqueue owner that is not a machineset")
+		return
+	}
+	c.controller.enqueueMachineSet(machineSet)
+}
+
+type jobFactory func(string) (*v1batch.Job, *kapi.ConfigMap, error)
+
+func (f jobFactory) BuildJob(name string) (*v1batch.Job, *kapi.ConfigMap, error) {
+	return f(name)
+}
+
+func (c *MachineSetController) getJobFactory(machineSet *clusteroperator.MachineSet) (controller.JobFactory, error) {
+	cluster, err := c.clusterForMachineSet(machineSet)
+	if err != nil {
+		return nil, err
+	}
+	return jobFactory(func(name string) (*v1batch.Job, *kapi.ConfigMap, error) {
+		vars, err := ansible.GenerateMachineSetVars(cluster, machineSet)
+		if err != nil {
+			return nil, nil, err
+		}
+		var playbook string
+		if machineSet.Spec.NodeType == clusteroperator.NodeTypeMaster {
+			playbook = masterProvisioningPlaybook
+		} else {
+			playbook = computeProvisioningPlaybook
+		}
+		job, configMap := c.ansibleGenerator.GeneratePlaybookJob(name, &cluster.Spec.Hardware, playbook, ansible.DefaultInventory, vars)
+		return job, configMap, nil
+	}), nil
+}
+
+func (c *MachineSetController) setMachineSetToNotProvisioning(original *clusteroperator.MachineSet) error {
+	machineSet := original.DeepCopy()
+	now := metav1.Now()
+
+	provisioning := machineSetCondition(machineSet, clusteroperator.MachineSetHardwareProvisioning)
+	if provisioning != nil &&
+		provisioning.Status == kapi.ConditionTrue {
+		provisioning.Status = kapi.ConditionFalse
+		provisioning.LastTransitionTime = now
+		provisioning.LastProbeTime = now
+		provisioning.Reason = "SpecChanged"
+		provisioning.Message = "Spec changed. New provisioning needed"
```

Comment:
Method could use a godoc, I'm a bit confused about the spec changed reason/message.